File: backend/src/main/java/com/brandonkamga/lescracks/dto/common/ApiError.java
package com.brandonkamga.lescracks.dto.common;

import com.brandonkamga.lescracks.exception.ErrorCode;
import com.fasterxml.jackson.annotation.JsonInclude;

import java.time.Instant;
import java.util.Map;

/**
 * What a failed request answers with.
 *
 * {@code code} is what a client branches on; {@code message} is what a person reads.
 * {@code reference} appears only on faults the caller cannot fix, and is the same string
 * printed beside the stack trace — a bug report quoting it points straight at the log line.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiError(
        ErrorCode code,
        String message,
        String path,
        Instant timestamp,
        /** Field name to message, on validation failures only. */
        Map<String, String> fields,
        String reference) {

    public static ApiError of(ErrorCode code, String message, String path) {
        return new ApiError(code, message, path, Instant.now(), null, null);
    }

    public static ApiError fields(String message, String path, Map<String, String> fields) {
        return new ApiError(ErrorCode.VALIDATION_FAILED, message, path, Instant.now(), fields, null);
    }

    public static ApiError referenced(ErrorCode code, String message, String path, String reference) {
        return new ApiError(code, message, path, Instant.now(), null, reference);
    }
}

File: frontend/src/lib/apiError.ts
/** Mirrors the ErrorCode enum on the API. */
export type ApiErrorCode =
  | 'NOT_FOUND'
  | 'BAD_REQUEST'
  | 'VALIDATION_FAILED'
  | 'FORBIDDEN'
  | 'UNAUTHENTICATED'
  | 'CONFLICT'
  | 'DATA_CONFLICT'
  | 'PAYLOAD_TOO_LARGE'
  | 'MALFORMED_REQUEST'
  | 'METHOD_NOT_ALLOWED'
  | 'INTERNAL_ERROR';

/**
 * A failed request, with everything the API said about it.
 *
 * Throwing a bare Error kept only the sentence, so a caller wanting to react — send the user
 * back to sign-in, highlight the field at fault, show the reference — had to match on text.
 */
export class ApiError extends Error {
  readonly status: number;
  readonly code?: ApiErrorCode;
  /** Quote this in a bug report; it is printed beside the stack trace server-side. */
  readonly reference?: string;
  /** Field name to message, present when code is VALIDATION_FAILED. */
  readonly fieldErrors?: Record<string, string>;

  constructor(
    message: string,
    status: number,
    code?: ApiErrorCode,
    reference?: string,
    fieldErrors?: Record<string, string>,
  ) {
    super(message);
    this.name = 'ApiError';
    this.status = status;
    this.code = code;
    this.reference = reference;
    this.fieldErrors = fieldErrors;
  }

  /** True when signing in again is what would fix it. */
  get needsSignIn(): boolean {
    return this.code === 'UNAUTHENTICATED' || this.status === 401;
  }

  /** True when the user cannot act on it and should be given the reference. */
  get isOurFault(): boolean {
    return this.code === 'INTERNAL_ERROR' || this.status >= 500;
  }
}

type ErrorBody = {
  message?: string;
  errorCode?: ApiErrorCode;
  reference?: string;
  data?: Record<string, string>;
};

/** Builds the error from a response body, whatever shape the server managed to return. */
export function apiErrorFrom(body: ErrorBody | null, status: number, fallback: string): ApiError {
  return new ApiError(
    body?.message || fallback,
    status,
    body?.errorCode,
    body?.reference,
    body?.errorCode === 'VALIDATION_FAILED' ? body?.data : undefined,
  );
}
